### PR TITLE
Merge ProgPow on top of  ethminer master

### PR DESCRIPTION
**Pull request by hackmod**
This is a test merge on top of the master branch of the ethminer.

- some fix unintentionally reverted (705c18621, 2a3fb4a, 68f0743e1d9) need to apply again carefully.
- ~~`set CUDA_HOST_COMPILER=/usr/bin/gcc-5` for cuda-8~~
- misc fixes to pass travis
- OpenCL only binary build added by default
- use CUDA-9.2 for OSX (to fix compile issue)

Thanks for your great job!

### CODE REVIEW

**Comment by AndreaLanfranchi** [comment]
What's going on here ?

**Comment by hackmod** [comment]
now all Travis ci tests passed! https://travis-ci.org/EthersocialNetwork/ethminer

**Comment by AndreaLanfranchi** [comment]
Can you please open a branch only focused on CI ?
There are a lot of commits by @jean-m-cyr which I believe may conflict with #61

**Comment by jean-m-cyr** [comment]
> some fix unintentionally reverted (705c186, 2a3fb4a, 68f0743) need to apply again carefully.

I would not be concerned with those. The --noexit option yields suboptimal results when enabled, it was more intended as a demonstration. The overlap scheduling is only advantageous when running a single stream on NVidia... nobody does that!

**Comment by hackmod** [comment]
> @AndreaLanfranchi wrote:
> Can you please open a branch only focused on CI ?
> There are a lot of commits by @jean-m-cyr which I believe may conflict with #61

as you can see travis/appveyor fixes has been applied after merging, and it seems less work needed.

> @jean-m-cyr wrote:
> > some fix unintentionally reverted ([705c186](https://github.com/AndreaLanfranchi/ethminer/commit/705c18621f3612ebaa26a45a776e766a5b5cb9da), [2a3fb4a](https://github.com/AndreaLanfranchi/ethminer/commit/2a3fb4a235d2e1031ab7a438f34846b580fcdca4), [68f0743](https://github.com/AndreaLanfranchi/ethminer/commit/68f0743e1d95b6ebdeb69121c06be633b1ae59d6)) need to apply again carefully.
> 
> I would not be concerned with those. The --noexit option yields suboptimal results when enabled, it was more intended as a demonstration. The overlap scheduling is only advantageous when running a single stream on NVidia... nobody does that!

Good to hear that!

It seems no significiant issue I can see after merging on the ethminer master branch, but PR #61 has many binaries, so only @jean-m-cyr can manage any possible conflicts easily.

How about try to merge it except PR #61 on to the ethminer master first ?(or develop branch)

After merge PR #61 on top of this PR(a64d39d), merge conflicts seems not so easy but almost all conflicts/unmergables are came from binaries and other fixes are seems rather small.
~~~
...
$ git merge refs/pull/61/head
warning: Cannot merge binary files: libethash-cl/kernels/bin/ethash_tonga_lws64.bin (HEAD vs. refs/pull/61/head)
warning: Cannot merge binary files: libethash-cl/kernels/bin/ethash_tonga_lws256.bin (HEAD vs. refs/pull/61/head)
warning: Cannot merge binary files: libethash-cl/kernels/bin/ethash_tonga_lws128.bin (HEAD vs. refs/pull/61/head)
...(snip)
Auto-merging libethash-cl/kernels/Makefile
Auto-merging libethash-cl/ethash_miner_kernel.cl
Auto-merging libethash-cl/CLMiner.cpp
Automatic merge failed; fix conflicts and then commit the result.
$ git reflog
a64d39d HEAD@{0}: checkout: moving from merge-progpow-master to merge-progpow-master
...(snip)
$ git diff --stat
 libethash-cl/kernels/bin/ethash_baffin_lws128.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_baffin_lws256.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_baffin_lws64.bin     | Unmerged
 libethash-cl/kernels/bin/ethash_ellesmere_lws128.bin | Unmerged
 libethash-cl/kernels/bin/ethash_ellesmere_lws256.bin | Unmerged
 libethash-cl/kernels/bin/ethash_ellesmere_lws64.bin  | Unmerged
 libethash-cl/kernels/bin/ethash_gfx900_lws128.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_gfx900_lws256.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_gfx900_lws64.bin     | Unmerged
 libethash-cl/kernels/bin/ethash_gfx901_lws128.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_gfx901_lws256.bin    | Unmerged
 libethash-cl/kernels/bin/ethash_gfx901_lws64.bin     | Unmerged
 libethash-cl/kernels/bin/ethash_tonga_lws128.bin     | Unmerged
 libethash-cl/kernels/bin/ethash_tonga_lws256.bin     | Unmerged
 libethash-cl/kernels/bin/ethash_tonga_lws64.bin      | Unmerged
 libethash-cl/kernels/isa/GCN3_ethash_search.isa      | Unmerged
 libethash-cl/kernels/isa/GCN3_ethash_search.isa      |  90 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----------------------
 libethash-cl/kernels/isa/GCN5_ethash_search.isa      | Unmerged
 libethash-cl/kernels/isa/GCN5_ethash_search.isa      | 100 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------
 2 files changed, 136 insertions(+), 54 deletions(-)
~~~

**Comment by jean-m-cyr** [comment]
You'll need to run the makefile in libethashcl/kernels to recreate the binaries that add then to the merge commit.

**Comment by hackmod** [comment]
> You'll need to run the makefile in libethashcl/kernels to recreate the binaries that add then to the merge commit.

good to hear! so then only two files have to be fixed
~~~
$ git diff | grep ^diff | grep isa
diff --cc libethash-cl/kernels/isa/GCN3_ethash_search.isa
diff --cc libethash-cl/kernels/isa/GCN5_ethash_search.isa
~~~

**Comment by hackmod** [comment]
I've just tried to merge/resolve conflicts PR #16 without any GCN/isa knowledge.
travis/appveyor builds successfully, but binaries do not work correctly.
https://travis-ci.org/EthersocialNetwork/ethminer/builds/488070169

anyway, I think @jean-m-cyr can manage it.

@AndreaLanfranchi // how do you think of the current progpow integration stage? It seems stable enough,  at least stable than reference work done by @ifdefelse or its forks.

**Comment by AndreaLanfranchi** [comment]
@hackmod 
I believe original @ifdefelse work was meant to be more a proof of concept than a fully fledged miner implementation. Personally I never got it to work properly **but** I must say I added a lot of features to help benchmark/debug which were not present in the ethminer version they forked from.

Actually this implementation is to be considered stable.

**Comment by hackmod** [comment]
stashed PR #61 cherry-picked manually (`*.isa` are overwritten), and ethash binaries now work!
(https://github.com/EthersocialNetwork/ethminer/releases/tag/v0.9.2-progpow-merge)

**Comment by jean-m-cyr** [comment]
PR 61 has issues. See PR notes.

**Comment by hackmod** [comment]
c382828 cherry-picked from PR #67 now kernel dmesg errors disappeared!

File: ethminer/main.cpp [hunk 1]
```diff
@@ -366,6 +368,12 @@ class MinerCLI
 
         app.add_option("--cpu-devices,--cp-devices", m_CPSettings.devices, "");
 
+#endif
+
+#if ETH_ETHASHCPU
```
Comment (AndreaLanfranchi) [inline ethminer/main.cpp]: `ETH_ETHASH` prefix has been removed from directives. It should be `_CPU` only

File: ethminer/main.cpp [hunk 2]
```diff
@@ -1013,6 +1022,20 @@ class MinerCLI
                  << endl;
         }
 
+        if (ctx == "cp")
```
Comment (AndreaLanfranchi) [inline ethminer/main.cpp]: This whole block is already present into master. Why it's being added once more ?

File: libdevcore/CommonData.h [hunk 3]
```diff
@@ -166,7 +166,7 @@ inline std::string toHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd)
     return (prefix == HexPrefix::Add) ? "0x" + str : str;
 }
 
-inline std::string toHex(uint64_t _n, HexPrefix _prefix = HexPrefix::DontAdd, short _bytes = 16)
+inline std::string toHex(uint64_t _n, HexPrefix _prefix = HexPrefix::DontAdd, int _bytes = 16)
```
Comment (AndreaLanfranchi) [inline libdevcore/CommonData.h]: In master there is already the change from short to int.

Comment (hackmod) [reply libdevcore/CommonData.h]: github's merge commit doesn't show correct `git show` result. this fix done by 8b42ec37

File: ethminer/main.cpp (hunk 2 above)
Comment (hackmod) [reply ethminer/main.cpp]: thanks! fixed

File: ethminer/main.cpp (hunk 1 above)
Comment (hackmod) [reply ethminer/main.cpp]: oops! fixed now

**Pull request closed without merging**